SimTracker/TrackTriggerAssociation/plugins/TTTrackAssociator.cc

```diff
@@ -17,62 +17,57 @@ void TTTrackAssociator<Ref_Phase2TrackerDigi_>::produce(edm::Event& iEvent, cons
 
     /// Get the Tracks already stored away
     edm::Handle<std::vector<TTTrack<Ref_Phase2TrackerDigi_>>> TTTrackHandle;
-    iEvent.getByToken(*iTag, TTTrackHandle);
+    iEvent.getByToken(iTag, TTTrackHandle);
 
     /// Prepare the necessary maps
-    std::map<edm::Ptr<TTTrack<Ref_Phase2TrackerDigi_>>, edm::Ptr<TrackingParticle>> trackToTrackingParticleMap;
-    std::map<edm::Ptr<TrackingParticle>, std::vector<edm::Ptr<TTTrack<Ref_Phase2TrackerDigi_>>>>
-        trackingParticleToTrackVectorMap;
+    std::map<TTTrackPtr, TrackingParticlePtr> trackToTrackingParticleMap;
+    std::map<TrackingParticlePtr, std::vector<TTTrackPtr>> trackingParticleToTrackVectorMap;
     trackToTrackingParticleMap.clear();
     trackingParticleToTrackVectorMap.clear();
 
     // Start the loop on tracks
 
-    unsigned int j = 0;  /// Counter needed to build the edm::Ptr to the TTTrack
-    typename std::vector<TTTrack<Ref_Phase2TrackerDigi_>>::const_iterator inputIter;
-    for (inputIter = TTTrackHandle->begin(); inputIter != TTTrackHandle->end(); ++inputIter) {
+    for (unsigned int jTrk = 0; jTrk < TTTrackHandle->size(); jTrk++) {
       /// Make the pointer to be put in the map
-      edm::Ptr<TTTrack<Ref_Phase2TrackerDigi_>> tempTrackPtr(TTTrackHandle, j++);
+      TTTrackPtr tempTrackPtr(TTTrackHandle, jTrk);
 
-      /// Get the stubs of the TTTrack (theseStubs)
-      std::vector<edm::Ref<edmNew::DetSetVector<TTStub<Ref_Phase2TrackerDigi_>>, TTStub<Ref_Phase2TrackerDigi_>>>
-          theseStubs = tempTrackPtr->getStubRefs();
+      /// Get all the stubs of the TTTrack (theseStubs)
+      std::vector<TTStubRef> theseStubs = tempTrackPtr->getStubRefs();
 
-      /// Auxiliary map to store TP addresses and TP edm::Ptr
-      std::map<const TrackingParticle*, edm::Ptr<TrackingParticle>> auxMap;
+      /// Auxiliary map to relate TP addresses and TP edm::Ptr
+      std::map<const TrackingParticle*, TrackingParticlePtr> auxMap;
       auxMap.clear();
       int mayCombinUnknown = 0;
 
-      /// Fill the inclusive map which is careless of the stub classification
-      for (unsigned int is = 0; is < theseStubs.size(); is++) {
-        //        std::vector< edm::Ref< edmNew::DetSetVector< TTCluster< Ref_Phase2TrackerDigi_ > >, TTCluster< Ref_Phase2TrackerDigi_ > > > theseClusters = theseStubs.at(is)->getClusterRefs();
+      /// Fill the map associating each TP to a vector of L1 tracks.
+      /// Do this using the association map of the clusters inside each stub,
+      /// as stub associator misses stub --> all TP map (FIX).
+      for (const TTStubRef& stub : theseStubs) {
         for (unsigned int ic = 0; ic < 2; ic++) {
-          std::vector<edm::Ptr<TrackingParticle>> tempTPs =
-              TTClusterAssociationMapHandle->findTrackingParticlePtrs(theseStubs.at(is)->clusterRef(ic));
-          for (unsigned int itp = 0; itp < tempTPs.size(); itp++)  // List of TPs linked to stub clusters
+          std::vector<TrackingParticlePtr> tempTPs =
```

Comment:
capture by const ref

Reply:
Can't as ClusterAssociationMap::findTrackingParticlePtrs(cluster) returns temporary vector if cluster is not in map.